Push more of the LLM conversation agent loop into ChatSession

allenporter opened the pull request:
<!--
  You are amazing! Thanks for contributing to our project!
  Please, DO NOT DELETE ANY TEXT from this template! (unless instructed).
-->

## Proposed change
<!--
  Describe the big picture of your changes here to communicate to the
  maintainers why we should accept this pull request. If it fixes a bug
  or resolves a feature request, be sure to link to that issue in the
  additional information section.
-->
Push more of the LLM conversation agent loop into ChatSession. This flow is the same for all conversation agents. This pulls another 40 lines of code out to a common library.

## Type of change
<!--
  What type of change does your PR introduce to Home Assistant?
  NOTE: Please, check only 1! box!
  If your PR requires multiple boxes to be checked, you'll most likely need to
  split it into multiple PRs. This makes things easier and faster to code review.
-->

- [ ] Dependency upgrade
- [ ] Bugfix (non-breaking change which fixes an issue)
- [ ] New integration (thank you!)
- [ ] New feature (which adds functionality to an existing integration)
- [ ] Deprecation (breaking change to happen in the future)
- [ ] Breaking change (fix/feature causing existing functionality to break)
- [x] Code quality improvements to existing code or addition of tests

## Additional information
<!--
  Details are important, and help maintainers processing your PR.
  Please be sure to fill out additional details, if applicable.
-->

- This PR fixes or closes issue: fixes #
- This PR is related to issue: 
- Link to documentation pull request: 

## Checklist
<!--
  Put an `x` in the boxes that apply. You can also fill these out after
  creating the PR. If you're unsure about any of them, don't hesitate to ask.
  We're here to help! This is simply a reminder of what we are going to look
  for before merging your code.
-->

- [ ] The code change is tested and works locally.
- [ ] Local tests pass. **Your PR cannot be merged unless tests pass**
- [ ] There is no commented out code in this PR.
- [ ] I have followed the [development checklist][dev-checklist]
- [ ] I have followed the [perfect PR recommendations][perfect-pr]
- [ ] The code has been formatted using Ruff (`ruff format homeassistant tests`)
- [ ] Tests have been added to verify that the new code works.

If user exposed functionality or configuration variables are added/changed:

- [ ] Documentation added/updated for [www.home-assistant.io][docs-repository]

If the code communicates with devices, web services, or third-party tools:

- [ ] The [manifest file][manifest-docs] has all fields filled out correctly.  
      Updated and included derived files by running: `python3 -m script.hassfest`.
- [ ] New or updated dependencies have been added to `requirements_all.txt`.  
      Updated by running `python3 -m script.gen_requirements_all`.
- [ ] For the updated dependencies - a link to the changelog, or at minimum a diff between library versions is added to the PR description.

<!--
  This project is very active and we have a high turnover of pull requests.

  Unfortunately, the number of incoming pull requests is higher than what our
  reviewers can review and merge so there is a long backlog of pull requests
  waiting for review. You can help here!
  
  By reviewing another pull request, you will help raise the code quality of
  that pull request and the final review will be faster. This way the general
  pace of pull request reviews will go up and your wait time will go down.
  
  When picking a pull request to review, try to choose one that hasn't yet
  been reviewed.

  Thanks for helping out!
-->

To help with the load of incoming pull requests:

- [ ] I have reviewed two other [open pull requests][prs] in this repository.

[prs]: https://github.com/home-assistant/core/pulls?q=is%3Aopen+is%3Apr+-author%3A%40me+-draft%3Atrue+-label%3Awaiting-for-upstream+sort%3Acreated-desc+review%3Anone+-status%3Afailure

<!--
  Thank you for contributing <3

  Below, some useful links you could explore:
-->
[dev-checklist]: https://developers.home-assistant.io/docs/development_checklist/
[manifest-docs]: https://developers.home-assistant.io/docs/creating_integration_manifest/
[quality-scale]: https://developers.home-assistant.io/docs/integration_quality_scale_index/
[docs-repository]: https://github.com/home-assistant/home-assistant.io
[perfect-pr]: https://developers.home-assistant.io/docs/review-process/#creating-the-perfect-pr

allenporter committed:
Push more of the LLM conversation agent loop into ChatSession

allenporter committed:
Push more of the LLM conversation agent loop into ChatSession

joostlek committed:
Bump aiowithings to 3.1.5 (#136350)

piitaya committed:
Update frontend to 20250109.2 (#136348)

astrandb committed:
Bump myuplink lib to 0.7.0 (#136343)

astrandb committed:
Avoid keyerror on incomplete api data in myuplink (#136333)

* Avoid keyerror

* Inject erroneous value in device point fixture

* Update diagnostics snapshot

jbouwh committed:
Improve error handling for incomfort gateway (#136317)

epenet committed:
Enable strict-typing in lovelace (#136327)

NoRi2909 committed:
Fix sentence-casing in isy994 integration strings, reword "lock user code" (#136316)

epenet committed:
Move single-use lovelace function (#136336)

epenet committed:
Use runtime_data in bosch_shc (#136356)

epenet committed:
Use runtime_data in canary (#136357)

NoRi2909 committed:
Several fixes in user-facing strings of Renson integration actions (#136279)

epenet committed:
Cleanup hass.data in cloudflare (#136358)

catsmanac committed:
Bump pyenphase to 1.23.1 (#136200)

firstof9 committed:
Properly parse AirNow API data in coordinator (#136198)

jbouwh committed:
Enable strict typing for incomfort integration (#136291)

* Enable strict typing for incomfort integration

* Comply to strict typing

* Wrap in bool

mattdoran committed:
Update Hydrawise maximum watering duration to meet the app limits (#136050)

Co-authored-by: Robert Resch <robert@resch.dev>

sdb9696 committed:
Fix tplink deprecated entity cleanup (#136160)

bdraco committed:
Refactor modbus polling to prevent dupe updates and memory leak (#136211)

andrewsayre committed:
Incorporate ControllerManager into HEOS Coordinator (#136302)

* Integrate ControllerManager

* Test for uncovered

* Correct test docstring

* Cast entry before graph access

* Assert config_entry state in reauth

* Use implicit casting

lanthaler committed:
Set Netgear device entities to unavailable when the device is not connected (#135362)

silamon committed:
Handle LinkPlay devices with no mac (#136272)

Co-authored-by: J. Nick Koston <nick@koston.org>

klaasnicolaas committed:
Bump powerfox to v1.2.1 (#136366)

bors-ltd committed:
holiday: asynchronously generate the entity name (#136354)

Asking the country translation was trigerring Babel to open a file, and
thus a blocking I/O.

crug80 committed:
Fix slave id equal to 0 (#136263)

Co-authored-by: J. Nick Koston <nick@koston.org>

bdraco committed:
Migrate modbus to use HassKey (#136379)

synesthesiam committed:
Don't translate state names in default agent responses (#136382)

Don't translate state names in responses

balloob committed:
Allow LLMs to get calendar events from exposed calendars (#136304)

epenet committed:
Use runtime_data in coinbase (#136381)

epenet committed:
Use runtime_data in daikin (#136376)

epenet committed:
Use runtime_data in comelit (#136384)

epenet committed:
Use runtime_data in ccm15 (#136378)

thecode committed:
Add translated action exceptions to LG webOS TV (#136397)

* Add translated action exceptions to LG webOS TV

* Apply suggestions from code review

Co-authored-by: Joost Lekkerkerker <joostlek@outlook.com>

---------

Co-authored-by: Joost Lekkerkerker <joostlek@outlook.com>

allenporter committed:
Fix native tool call response handling

allenporter committed:
Fix native tool call response handling

allenporter committed:
Factor out async_call_tools into a separate method

allenporter committed:
Factor out async_call_tools into a separate method

allenporter committed:
Move some of the message flow back into the handler

allenporter committed:
Move some of the message flow back into the handler

dependabot[bot] committed:
Bump github/codeql-action from 3.28.3 to 3.28.4 (#136401)

Co-authored-by: dependabot[bot] <49699333+dependabot[bot]@users.noreply.github.com>

dependabot[bot] committed:
Bump codecov/codecov-action from 5.2.0 to 5.3.0 (#136402)

Co-authored-by: dependabot[bot] <49699333+dependabot[bot]@users.noreply.github.com>

gjohansson-ST committed:
Remove deprecated 17track package sensor (#136389)

Makrit committed:
Handle width and height placeholders in the thumbnail URL (#136227)

epenet committed:
Use runtime_data in coolmaster (#136405)

* Use runtime_data in coolmaster

* Adjust test

bdraco committed:
Bump aioharmony to 0.4.0 (#136398)

Thomas55555 committed:
Bump aioautomower to 2025.1.1 (#136365)

bdraco committed:
Bump WSDiscovery to 2.1.2 (#136363)

epenet committed:
Move denonavr shared constants to central location (#136421)

epenet committed:
Simplify update listener in denonavr (#136422)

frenck committed:
Omit Peblar update entities for most white label devices (#136374)

epenet committed:
Use runtime_data in denonavr (#136424)

VandeurenGlenn committed:
Bump nhc to 0.3.9 (#136418)

iprak committed:
Fix humidifier mode for Vesync (#135746)

Co-authored-by: Joost Lekkerkerker <joostlek@outlook.com>

andrewsayre committed:
Incorporate SourceManager into HEOS Coordinator (#136377)

Co-authored-by: Joost Lekkerkerker <joostlek@outlook.com>

jbouwh committed:
Rename incomfort exceptions classes to fix typo and assign correct translation domain (#136426)

erwindouna committed:
Tado change to async and add Data Update Coordinator (#134175)

Co-authored-by: Joost Lekkerkerker <joostlek@outlook.com>

thecode committed:
Add missing translations for LG webOS TV and fix names (#136438)

dknowles2 committed:
Refactor the Hydrawise config flow (#135886)

Co-authored-by: Joost Lekkerkerker <joostlek@outlook.com>

epenet committed:
Move dormakaba_dkey coordinator to separate module (#136437)

epenet committed:
Use runtime_data in directv (#136435)

epenet committed:
Move dexcom coordinator to separate module (#136433)

epenet committed:
Use runtime_data in dormakaba_dkey (#136440)

epenet committed:
Use runtime_data in duotecno (#136444)

epenet committed:
Use runtime_data in dexcom (#136441)

bdraco committed:
Bump aioharmony to 0.4.1 (#136413)

changelog: https://github.com/Harmony-Libs/aioharmony/compare/v0.4.0...v0.4.1

epenet committed:
Use runtime_data in dunehd (#136443)

arturpragacz committed:
Reorganize input sources in Onkyo options (#133511)

epenet committed:
Move dynalite service definitions to separate module (#136446)

Taraman17 committed:
Add more sensors to homee (#136445)

zweckj committed:
Bump aioacaia to 0.1.14 (#136453)

sdb9696 committed:
Add PARALLEL_UPDATES constant to ring integration platforms (#136470)

RaHehl committed:
Bump uiprotect to version 7.5.0 (#136475)

NoRi2909 committed:
Fix sentence-casing in strings of Vizio integration (#136465)

joostlek committed:
Make Spotify polling interval dynamic (#136461)

choumarin committed:
Powerwall: Reuse authentication cookie (#136147)

Co-authored-by: J. Nick Koston <nick@koston.org>

StevenLooman committed:
Bump async-upnp-client to 0.43.0 (#136481)

bdraco committed:
Bump pydantic to 2.10.6 (#136483)

epenet committed:
Cleanup update_listener in deconz (#136416)

epenet committed:
Move deconz function to util.py (#136414)

bdraco committed:
Fix httpx late import of trio doing blocking I/O in the event loop (#136409)

httpx 0.28.1 moved the trio import to happen a bit later

```
2025-01-23 19:53:12.370 WARNING (MainThread) [homeassistant.util.loop] Detected blocking call to open with args (/lib/c, rb) inside the event loop by integration rest at homeassistant/components/rest/data.py, line 88: self._async_client = create_async_httpx_client( (offender: /usr/local/lib/python3.13/ctypes/util.py, line 285: with open(filepath, rb) as fh:), please create a bug report at https://github.com/home-assistant/core/issues?q=is%3Aopen+is%3Aissue+label%3A%22integration%3A+rest%22
For developers, please see https://developers.home-assistant.io/docs/asyncio_blocking_operations/#open
Traceback (most recent call last):
  File "<frozen runpy>", line 198, in _run_module_as_main
  File "<frozen runpy>", line 88, in _run_code
  File "/usr/src/homeassistant/homeassistant/__main__.py", line 227, in <module>
    sys.exit(main())
  File "/usr/src/homeassistant/homeassistant/__main__.py", line 213, in main
    exit_code = runner.run(runtime_conf)
  File "/usr/src/homeassistant/homeassistant/runner.py", line 154, in run
    return loop.run_until_complete(setup_and_run_hass(runtime_config))
  File "/usr/local/lib/python3.13/asyncio/base_events.py", line 707, in run_until_complete
    self.run_forever()
  File "/usr/local/lib/python3.13/asyncio/base_events.py", line 678, in run_forever
    self._run_once()
  File "/usr/local/lib/python3.13/asyncio/base_events.py", line 2033, in _run_once
    handle._run()
  File "/usr/local/lib/python3.13/asyncio/events.py", line 89, in _run
    self._context.run(self._callback, *self._args)
  File "/usr/src/homeassistant/homeassistant/config_entries.py", line 2360, in _async_forward_entry_setup
    result = await async_setup_component(
  File "/usr/src/homeassistant/homeassistant/setup.py", line 165, in async_setup_component
    result = await _async_setup_component(hass, domain, config)
  File "/usr/src/homeassistant/homeassistant/setup.py", line 420, in _async_setup_component
    result = await task
  File "/usr/src/homeassistant/homeassistant/components/sensor/__init__.py", line 90, in async_setup
    await component.async_setup(config)
  File "/usr/src/homeassistant/homeassistant/helpers/entity_component.py", line 146, in async_setup
    self.hass.async_create_task_internal(
  File "/usr/src/homeassistant/homeassistant/core.py", line 832, in async_create_task_internal
    task = create_eager_task(target, name=name, loop=self.loop)
  File "/usr/src/homeassistant/homeassistant/util/async_.py", line 45, in create_eager_task
    return Task(coro, loop=loop, name=name, eager_start=True)
  File "/usr/src/homeassistant/homeassistant/helpers/entity_component.py", line 307, in async_setup_platform
    await self._platforms[key].async_setup(platform_config, discovery_info)
  File "/usr/src/homeassistant/homeassistant/helpers/entity_platform.py", line 303, in async_setup
    await self._async_setup_platform(async_create_setup_awaitable)
  File "/usr/src/homeassistant/homeassistant/helpers/entity_platform.py", line 363, in _async_setup_platform
    awaitable = create_eager_task(awaitable, loop=hass.loop)
  File "/usr/src/homeassistant/homeassistant/util/async_.py", line 45, in create_eager_task
    return Task(coro, loop=loop, name=name, eager_start=True)
  File "/usr/src/homeassistant/homeassistant/components/rest/sensor.py", line 85, in async_setup_platform
    await rest.async_update(log_errors=False)
  File "/usr/src/homeassistant/homeassistant/components/rest/data.py", line 88, in async_update
    self._async_client = create_async_httpx_client(

2025-01-23 19:53:12.371 WARNING (MainThread) [homeassistant.util.loop] Detected blocking call to glob with args (/lib/libc.so,) inside the event loop by integration rest at homeassistant/components/rest/data.py, line 88: self._async_client = create_async_httpx_client( (offender: /usr/local/lib/python3.13/ctypes/util.py, line 311: for f in glob({0}{1}.format(prefix, suffix)):), please create a bug report at https://github.com/home-assistant/core/issues?q=is%3Aopen+is%3Aissue+label%3A%22integration%3A+rest%22
For developers, please see https://developers.home-assistant.io/docs/asyncio_blocking_operations/#glob
Traceback (most recent call last):
  File "<frozen runpy>", line 198, in _run_module_as_main
  File "<frozen runpy>", line 88, in _run_code
  File "/usr/src/homeassistant/homeassistant/__main__.py", line 227, in <module>
    sys.exit(main())
  File "/usr/src/homeassistant/homeassistant/__main__.py", line 213, in main
    exit_code = runner.run(runtime_conf)
  File "/usr/src/homeassistant/homeassistant/runner.py", line 154, in run
    return loop.run_until_complete(setup_and_run_hass(runtime_config))
  File "/usr/local/lib/python3.13/asyncio/base_events.py", line 707, in run_until_complete
    self.run_forever()
  File "/usr/local/lib/python3.13/asyncio/base_events.py", line 678, in run_forever
    self._run_once()
  File "/usr/local/lib/python3.13/asyncio/base_events.py", line 2033, in _run_once
    handle._run()
  File "/usr/local/lib/python3.13/asyncio/events.py", line 89, in _run
    self._context.run(self._callback, *self._args)
  File "/usr/src/homeassistant/homeassistant/config_entries.py", line 2360, in _async_forward_entry_setup
    result = await async_setup_component(
  File "/usr/src/homeassistant/homeassistant/setup.py", line 165, in async_setup_component
    result = await _async_setup_component(hass, domain, config)
  File "/usr/src/homeassistant/homeassistant/setup.py", line 420, in _async_setup_component
    result = await task
  File "/usr/src/homeassistant/homeassistant/components/sensor/__init__.py", line 90, in async_setup
    await component.async_setup(config)
  File "/usr/src/homeassistant/homeassistant/helpers/entity_component.py", line 146, in async_setup
    self.hass.async_create_task_internal(
  File "/usr/src/homeassistant/homeassistant/core.py", line 832, in async_create_task_internal
    task = create_eager_task(target, name=name, loop=self.loop)
  File "/usr/src/homeassistant/homeassistant/util/async_.py", line 45, in create_eager_task
    return Task(coro, loop=loop, name=name, eager_start=True)
  File "/usr/src/homeassistant/homeassistant/helpers/entity_component.py", line 307, in async_setup_platform
    await self._platforms[key].async_setup(platform_config, discovery_info)
  File "/usr/src/homeassistant/homeassistant/helpers/entity_platform.py", line 303, in async_setup
    await self._async_setup_platform(async_create_setup_awaitable)
  File "/usr/src/homeassistant/homeassistant/helpers/entity_platform.py", line 363, in _async_setup_platform
    awaitable = create_eager_task(awaitable, loop=hass.loop)
  File "/usr/src/homeassistant/homeassistant/util/async_.py", line 45, in create_eager_task
    return Task(coro, loop=loop, name=name, eager_start=True)
  File "/usr/src/homeassistant/homeassistant/components/rest/sensor.py", line 85, in async_setup_platform
    await rest.async_update(log_errors=False)
  File "/usr/src/homeassistant/homeassistant/components/rest/data.py", line 88, in async_update
    self._async_client = create_async_httpx_client(

2025-01-23 19:53:12.372 WARNING (MainThread) [homeassistant.util.loop] Detected blocking call to iglob with args (/lib/libc.so,) inside the event loop by integration rest at homeassistant/components/rest/data.py, line 88: self._async_client = create_async_httpx_client( (offender: /usr/local/lib/python3.13/glob.py, line 31: return list(iglob(pathname, root_dir=root_dir, dir_fd=dir_fd, recursive=recursive,), please create a bug report at https://github.com/home-assistant/core/issues?q=is%3Aopen+is%3Aissue+label%3A%22integration%3A+rest%22
For developers, please see https://developers.home-assistant.io/docs/asyncio_blocking_operations/#iglob
Traceback (most recent call last):
  File "<frozen runpy>", line 198, in _run_module_as_main
  File "<frozen runpy>", line 88, in _run_code
  File "/usr/src/homeassistant/homeassistant/__main__.py", line 227, in <module>
    sys.exit(main())
  File "/usr/src/homeassistant/homeassistant/__main__.py", line 213, in main
    exit_code = runner.run(runtime_conf)
  File "/usr/src/homeassistant/homeassistant/runner.py", line 154, in run
    return loop.run_until_complete(setup_and_run_hass(runtime_config))
  File "/usr/local/lib/python3.13/asyncio/base_events.py", line 707, in run_until_complete
    self.run_forever()
  File "/usr/local/lib/python3.13/asyncio/base_events.py", line 678, in run_forever
    self._run_once()
  File "/usr/local/lib/python3.13/asyncio/base_events.py", line 2033, in _run_once
    handle._run()
  File "/usr/local/lib/python3.13/asyncio/events.py", line 89, in _run
    self._context.run(self._callback, *self._args)
  File "/usr/src/homeassistant/homeassistant/config_entries.py", line 2360, in _async_forward_entry_setup
    result = await async_setup_component(
  File "/usr/src/homeassistant/homeassistant/setup.py", line 165, in async_setup_component
    result = await _async_setup_component(hass, domain, config)
  File "/usr/src/homeassistant/homeassistant/setup.py", line 420, in _async_setup_component
    result = await task
  File "/usr/src/homeassistant/homeassistant/components/sensor/__init__.py", line 90, in async_setup
    await component.async_setup(config)
  File "/usr/src/homeassistant/homeassistant/helpers/entity_component.py", line 146, in async_setup
    self.hass.async_create_task_internal(
  File "/usr/src/homeassistant/homeassistant/core.py", line 832, in async_create_task_internal
    task = create_eager_task(target, name=name, loop=self.loop)
  File "/usr/src/homeassistant/homeassistant/util/async_.py", line 45, in create_eager_task
    return Task(coro, loop=loop, name=name, eager_start=True)
  File "/usr/src/homeassistant/homeassistant/helpers/entity_component.py", line 307, in async_setup_platform
    await self._platforms[key].async_setup(platform_config, discovery_info)
  File "/usr/src/homeassistant/homeassistant/helpers/entity_platform.py", line 303, in async_setup
    await self._async_setup_platform(async_create_setup_awaitable)
  File "/usr/src/homeassistant/homeassistant/helpers/entity_platform.py", line 363, in _async_setup_platform
    awaitable = create_eager_task(awaitable, loop=hass.loop)
  File "/usr/src/homeassistant/homeassistant/util/async_.py", line 45, in create_eager_task
    return Task(coro, loop=loop, name=name, eager_start=True)
  File "/usr/src/homeassistant/homeassistant/components/rest/sensor.py", line 85, in async_setup_platform
    await rest.async_update(log_errors=False)
  File "/usr/src/homeassistant/homeassistant/components/rest/data.py", line 88, in async_update
    self._async_client = create_async_httpx_client(

2025-01-23 19:53:12.374 WARNING (MainThread) [homeassistant.util.loop] Detected blocking call to scandir with args (/lib,) inside the event loop by integration rest at homeassistant/components/rest/data.py, line 88: self._async_client = create_async_httpx_client( (offender: /usr/local/lib/python3.13/glob.py, line 170: with os.scandir(arg) as it:), please create a bug report at https://github.com/home-assistant/core/issues?q=is%3Aopen+is%3Aissue+label%3A%22integration%3A+rest%22
For developers, please see https://developers.home-assistant.io/docs/asyncio_blocking_operations/#scandir
Traceback (most recent call last):
  File "<frozen runpy>", line 198, in _run_module_as_main
  File "<frozen runpy>", line 88, in _run_code
  File "/usr/src/homeassistant/homeassistant/__main__.py", line 227, in <module>
    sys.exit(main())
  File "/usr/src/homeassistant/homeassistant/__main__.py", line 213, in main
    exit_code = runner.run(runtime_conf)
  File "/usr/src/homeassistant/homeassistant/runner.py", line 154, in run
    return loop.run_until_complete(setup_and_run_hass(runtime_config))
  File "/usr/local/lib/python3.13/asyncio/base_events.py", line 707, in run_until_complete
    self.run_forever()
  File "/usr/local/lib/python3.13/asyncio/base_events.py", line 678, in run_forever
    self._run_once()
  File "/usr/local/lib/python3.13/asyncio/base_events.py", line 2033, in _run_once
    handle._run()
  File "/usr/local/lib/python3.13/asyncio/events.py", line 89, in _run
    self._context.run(self._callback, *self._args)
  File "/usr/src/homeassistant/homeassistant/config_entries.py", line 2360, in _async_forward_entry_setup
    result = await async_setup_component(
  File "/usr/src/homeassistant/homeassistant/setup.py", line 165, in async_setup_component
    result = await _async_setup_component(hass, domain, config)
  File "/usr/src/homeassistant/homeassistant/setup.py", line 420, in _async_setup_component
    result = await task
  File "/usr/src/homeassistant/homeassistant/components/sensor/__init__.py", line 90, in async_setup
    await component.async_setup(config)
  File "/usr/src/homeassistant/homeassistant/helpers/entity_component.py", line 146, in async_setup
    self.hass.async_create_task_internal(
  File "/usr/src/homeassistant/homeassistant/core.py", line 832, in async_create_task_internal
    task = create_eager_task(target, name=name, loop=self.loop)
  File "/usr/src/homeassistant/homeassistant/util/async_.py", line 45, in create_eager_task
    return Task(coro, loop=loop, name=name, eager_start=True)
  File "/usr/src/homeassistant/homeassistant/helpers/entity_component.py", line 307, in async_setup_platform
    await self._platforms[key].async_setup(platform_config, discovery_info)
  File "/usr/src/homeassistant/homeassistant/helpers/entity_platform.py", line 303, in async_setup
    await self._async_setup_platform(async_create_setup_awaitable)
  File "/usr/src/homeassistant/homeassistant/helpers/entity_platform.py", line 363, in _async_setup_platform
    awaitable = create_eager_task(awaitable, loop=hass.loop)
  File "/usr/src/homeassistant/homeassistant/util/async_.py", line 45, in create_eager_task
    return Task(coro, loop=loop, name=name, eager_start=True)
  File "/usr/src/homeassistant/homeassistant/components/rest/sensor.py", line 85, in async_setup_platform
    await rest.async_update(log_errors=False)
  File "/usr/src/homeassistant/homeassistant/components/rest/data.py", line 88, in async_update
    self._async_client = create_async_httpx_client(
```

tl-sl committed:
Bump pysmlight to v0.1.6 (#136496)

sdb9696 committed:
Update tplink climate platform to use thermostat module (#136166)

joostlek committed:
Refactor EZVIZ config flow tests (#136434)

starkillerOG committed:
Bump reolink_aio to 0.11.8 (#136504)

lboue committed:
Add TemperatureLevel feature from Matter TemperatureControl cluster (#134532)

tl-sl committed:
Abort config flow is device is unsupported (#136505)

* Abort config flow if device is not yet supported

* Abort on user step for unsupported device

* Add string for unsupported device

* fix tests due to extra get_info calls

* add tests for unsupported devices to abort flow

joostlek committed:
Fix Spotify flaky test (#136529)

andrewsayre committed:
Incorporate GroupManager into HEOS Coordinator (#136462)

* Incorporate GroupManager

* Update quality scale

* Fix group params

* Revert quality scale change

* Rename varaible

* Move group action implementaton out of coordinator

* Fix get_group_members hass access

* entity -> entity_id

bdraco committed:
Reduce boilerplate code to setup modbus platform entities (#136491)

mkmer committed:
Bump AIOSomeComfort to 0.0.30 in Honeywell (#136523)

abmantis committed:
Reuse fixtures in config flow tests for Whirlpool (#136489)

* Use fixtures in config flow tests for Whirlpool

* Keep old tests; new one will go to separate PR

iprak committed:
Bump pyvesync to 2.1.16 (#136493)

Update use pyvesync 2.1.16

Co-authored-by: Shay Levy <levyshay1@gmail.com>

bdraco committed:
Remove unneeded call active check in modbus (#136487)

We have an asyncio.Lock in place to prevent polling collisions
now so this is no longer needed

Co-authored-by: Joost Lekkerkerker <joostlek@outlook.com>
Co-authored-by: Shay Levy <levyshay1@gmail.com>

allenporter committed:
Attempt at a hacky async generator

allenporter committed:
Attempt at a hacky async generator

balloob commented on homeassistant/components/conversation/session.py:
```diff
@@ -326,3 +362,91 @@ async def async_update_llm_data(
+
+    async def _async_call_tools(
+        self,
+        agent_id: str | None,
```
We shouldn't have to pass in `agent_id`. I think that we should have a ChatSession with an active conversation with the ConversationInput that started this conversation. That input contains all the useful info.

keithle888 committed:
Updated igloohome-api dependency to 0.1.0 (#136516)

- Updated igloohome-api to 0.1.0

CFenner committed:
Add heat pump heating rod sensors in ViCare integration (#136467)

* add heating rod sensors

* add labels

* update snapshot

mib1185 committed:
Add backup agent to Synology DSM (#135227)

* pre-alpha state

* small type

* use ChunkAsyncStreamIterator from aiohttp_client helper

* create parent folders during upload if none exists

* check file station permissionsduring setup

* ensure backup-agents are reloaded

* adjust config flow

* fix check for availability of file station

* fix possible unbound

* add config flow tests

* fix existing tests

* add backup tests

* backup listeners are not async

* some more tests

* migrate existing config entries

* fix migration

* notify backup listeners only when needed during setup

* add backup settings to options flow

* switch back to the listener approach from the dev docs example

* add negative tests

* fix tests

* use HassKey

* fix tests

* Revert "use HassKey"

This reverts commit 71c5a4d6fa9c04b4907ff5f8df6ef7bd1737aa85.

* use hass loop call_soon instead of non-eager-start tasks

* use HassKey for backup-agent-listeners

* delete empty backup-agent-listener list from hass.data

* don't handle single file download errors

* Apply suggestions from code review

Co-authored-by: J. Nick Koston <nick@koston.org>

* add more tests

* we don't have entities related to file_station api

* add more backup tests

* test unload backup agent

* revert sorting of properties

* additional use hass config location for default backup path

---------

Co-authored-by: J. Nick Koston <nick@koston.org>

squishykid committed:
Add Darsstar as codeowner for solax integration  (#136528)

* Add Darsstar as codeowner for solax integration

* Update manifest.json

NoRi2909 committed:
Fix wrong plural on tado.add_meter_reading action (#136524)

As this action can only take a single argument the plural introduced in the descriptions is misleading.

This also makes the friendly name of the action consistent with its key name.

MartinHjelmare committed:
Add restore backup tests (#136538)

* Test restore backup with busy manager

* Test restore backup with agent error

* Test restore backup with file error

allenporter committed:
Simplify streaming tool loop

allenporter committed:
Simplify streaming tool loop

allenporter committed:
Revert exceptions

allenporter committed:
Revert exceptions

allenporter committed:
Revert some style changes

allenporter committed:
Revert some style changes

allenporter committed:
Reduce lines of code changed

allenporter committed:
Reduce lines of code changed

allenporter committed:
Remove tools function and re-order functions

allenporter committed:
Remove tools function and re-order functions

allenporter committed:
Revert tool call changes

allenporter committed:
Revert tool call changes

dan-r committed:
Bump ohmepy version to 1.2.6 (#136547)

allenporter committed:
Remove _NativeT typevar

allenporter committed:
Remove _NativeT typevar

allenporter committed:
Revert the streaming API changes

allenporter committed:
Revert the streaming API changes

allenporter committed:
Revert pre-commit

allenporter committed:
Revert pre-commit

allenporter committed:
Propagate HomeAssistantError as conversation result error message string

allenporter committed:
Propagate HomeAssistantError as conversation result error message string

allenporter committed:
Update async_get_messages to use input agent id

allenporter committed:
Update async_get_messages to use input agent id

allenporter committed:
Revert unnecessary deletion

allenporter committed:
Revert unnecessary deletion

allenporter committed:
Revert more unnecessary changes

allenporter committed:
Revert more unnecessary changes

allenporter committed:
Revert logging changes

allenporter committed:
Revert logging changes

allenporter committed:
Add tests for tool call methods

allenporter committed:
Add tests for tool call methods

allenporter committed:
Revert whitespace change

allenporter committed:
Revert whitespace change

andrewsayre committed:
Move action implementation out of HEOS Coordinator (#136539)

* Move play_source

* Update property docstring

* Correct import location

tronikos committed:
Bump opower to 0.8.8 (#136555)

Taraman17 committed:
Homee cover-test (#136563)

initial cover-test

dan-r committed:
Add select platform to Ohme (#136536)

* Add select platform

* Formatting

* Add parallel updates to select

* Remove comments

Taraman17 committed:
Bump pyHomee to 1.2.5 (#136567)

NoRi2909 committed:
Make casing of "server" and action descriptions consistent (#136561)

andrewsayre committed:
Use typed ConfigEntry throughout HEOS (#136569)

andrewsayre committed:
Bump HEOS quality scale to silver (#136533)

bump heos quality scale

andrewsayre committed:
Don't cast type in HEOS services (#136583)

andrewsayre committed:
Fix HEOS play media type playlist (#136585)

NoRi2909 committed:
Fix spelling of "Home Assistant" and "IDs" in xiaomi_aqara (#136578)

NoRi2909 committed:
Fix sentence-casing in action names, spelling of "IDs" (#136576)

epenet committed:
Fix optional argument in deconz test type definition (#136411)

joostlek committed:
Bump yt-dlp to 2025.01.26 (#136581)

balloob commented on homeassistant/components/conversation/agent_manager.py:
```diff
@@ -109,7 +110,19 @@ async def async_converse(
+                intent.IntentResponseErrorCode.UNKNOWN,
+                str(err),
+            )
+            return ConversationResult(
```
Let's make sure it's still tracked in the trace:
```suggestion
            result = ConversationResult(
```

allenporter replied:
Fixed, and added a test verifying it shows up in the trace.

balloob commented on homeassistant/components/openai_conversation/conversation.py:
```diff
@@ -94,6 +92,18 @@ def _message_convert(message: ChatCompletionMessage) -> ChatCompletionMessagePar
+    message: conversation.ChatMessage[ChatCompletionMessageParam],
+) -> ChatCompletionMessageParam:
+    """Convert a chat message to the native format."""
+    if message.native is not None:
```
You need to verify the agent_id is also correct. Otherwise, if Google adds a native message, OpenAI would use that object

allenporter replied:
Does `async_get_messages` do this filtering for us or am i missing some subtlety there?

balloob replied:
The filtering only applies to pure native messages. You can have a user message with a native object from another conversation agent

balloob reviewed (approved):
Great cleanup! 2 minor comments, otherwise let's goooo

jpelgrom committed:
Fix LetPot reauthentication flow tests setting up config entry (#136589)

Fix LetPot reauth tests setting up config entry

allenporter committed:
Add a test that explicit failure is traced

allenporter committed:
Add a test that explicit failure is traced

allenporter committed:
Add ChatSession message to filter agent messages

allenporter committed:
Merge branch 'session-api' of https://github.com/allenporter/home-assistant-core into session-api
